tools/shared/src/main/scala/org/scalajs/core/tools/linker/backend/emitter/IncClassEmitter.scala

```diff
@@ -0,0 +1,62 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2016, LAMP/EPFL        **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package org.scalajs.core.tools.linker.backend.emitter
+
+import scala.collection.mutable
+
+import org.scalajs.core.tools.sem._
+
+import org.scalajs.core.tools.linker.backend.OutputMode
+import org.scalajs.core.tools.linker.LinkingUnit
+
+/** Tracks incrementality breaking changes made during class emitting.
+ */
+final class IncClassEmitter(semantics: Semantics, outputMode: OutputMode) {
+
+  private var currentClassesCtorOpt: Set[String] = Set.empty
+  private[this] val askedForCtorOpt =
+    mutable.Map.empty[String, Set[(String, String, Boolean)]]
+
+  def beginRun(unit: LinkingUnit,
+      invalidateFunc: (String, String, Boolean) => Unit): ScalaJSClassEmitter = {
+    val classEmitter = new ScalaJSClassEmitter(outputMode, unit, this)
+    val newClassesCtorOpt = if (outputMode == OutputMode.ECMAScript6StrongMode) {
+      Set.empty[String]
+    } else {
+      classEmitter.candidateForJSConstructorOpt
+    }
+    val classesWhoChanged = (currentClassesCtorOpt -- newClassesCtorOpt) ++
+      (newClassesCtorOpt -- currentClassesCtorOpt)
+    currentClassesCtorOpt = newClassesCtorOpt
+
+    askedForCtorOpt.retain { (className, callers) =>
+      if (classesWhoChanged(className)) {
+        // invalidate
+        callers.foreach(invalidateFunc.tupled)
+        false
+      } else {
+        true
+      }
+    }
+    classEmitter
+  }
+
+  def usesJSConstructorOpt(className: String): Boolean = {
+    currentClassesCtorOpt(className)
+  }
+
+  def usesJSConstructorOpt(className: String, callerClassName: String,
+      methodName: String, isStatic: Boolean): Boolean = {
+    val currentEntry = askedForCtorOpt.getOrElse(className, Set.empty)
+    val newEntry = currentEntry + ((callerClassName, methodName, isStatic))
+    askedForCtorOpt += (className -> newEntry)
```

Comment:
Ah! That's where `getOrElseUpdate` comes to the rescue :-)
```scala
val currentEntry = askedForCtorOpt.getOrElseUpdate(className, mutable.Set.empty)
currentEntry += ((callerClassName, methodName, isStatic))
```
The `getOrElseUpdate` will install the newly created mutable set in `askedForCtorOpt` if `className` was not already there.